atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/sweep/queue/TargetedSweepTest.java

```diff
@@ -0,0 +1,273 @@
+        assertLatestEntryForCellInKvsAtTimestamp(TABLE_THOR, TEST_CELL, t2.getTimestamp());
+
+        verify(sweepQueue, times(2)).enqueue(anyList());
+    }
+
+    @Test
+    public void writesAddedToSweepQueueAndKvsOnPreCommitConditionFailure() {
+        long startTs = putWriteAndFailOnPreCommitConditionReturningStartTimestamp(SINGLE_WRITE);
+
+        assertThat(getEnqueuedWritesNumber(1)).containsExactly(WriteInfo.of(SINGLE_WRITE, startTs));
+        verify(sweepQueue, times(1)).enqueue(anyList());
+        assertLatestEntryForCellInKvsAtTimestamp(TABLE_CONS, TEST_CELL, startTs);
+    }
+
+    @Test
+    public void writesNotAddedToSweepQueueOrKvsOnWriteWriteConflict() {
+        Transaction t1 = txManager.createNewTransaction();
+        Transaction t2 = txManager.createNewTransaction();
+
+        put(t1, SINGLE_WRITE);
+        put(t2, SINGLE_WRITE);
+
+        t1.commit();
+        assertThatThrownBy(t2::commit).isInstanceOf(TransactionConflictException.class);
+
+        verify(sweepQueue, times(1)).enqueue(anyList());
+        assertLatestEntryForCellInKvsAtTimestamp(TABLE_CONS, TEST_CELL, t1.getTimestamp());
+    }
+
+    @Test
+    public void writesNotAddedToSweepQueueOrKvsOnException() {
+        assertThatThrownBy(() ->
+                txManager.runTaskWithRetry(txn -> {
+                    txn.put(TABLE_CONS, ImmutableMap.of(TEST_CELL, TEST_DATA));
+                    throw new RuntimeException("test");
+                })).isInstanceOf(RuntimeException.class);
+
+        verify(sweepQueue, times(0)).enqueue(anyList());
+        assertNoEntryForCellInKvs(TABLE_CONS, TEST_CELL);
+    }
+
+    @Test
+    public void sweepDeletesValuesWrittenOnPreCommitConditionFailure() {
+        useOneSweepQueueShard();
+        long startTs = putWriteAndFailOnPreCommitConditionReturningStartTimestamp(SINGLE_WRITE);
+
+        serializableTxManager.setUnreadableTimestamp(startTs + 1);
+        sweepQueue.sweepNextBatch(ShardAndStrategy.conservative(0));
+        assertNoEntryForCellInKvs(TABLE_CONS, TEST_CELL);
+    }
+
+    @Test
+    public void sweepRetainsLatestVersion() {
+        useOneSweepQueueShard();
+
+        writeInTransactionAndGetStartTimestamp(SINGLE_WRITE);
+        putWriteAndFailOnPreCommitConditionReturningStartTimestamp(SINGLE_WRITE);
+        long freshCommittedTs = writeInTransactionAndGetStartTimestamp(SINGLE_WRITE);
+        putWriteAndFailOnPreCommitConditionReturningStartTimestamp(SINGLE_DELETE);
+        long freshFailedTs = putWriteAndFailOnPreCommitConditionReturningStartTimestamp(SINGLE_WRITE);
+
+        serializableTxManager.setUnreadableTimestamp(freshFailedTs + 1);
+        sweepQueue.sweepNextBatch(ShardAndStrategy.conservative(0));
+        assertLatestEntryForCellInKvsAtTimestamp(TABLE_CONS, TEST_CELL, freshCommittedTs);
+        assertOnlySentinelBeforeTs(TABLE_CONS, TEST_CELL, freshCommittedTs);
+    }
```

Comment:
Will this test fail if the `SINGLE_WRITE` happens at a timestamp just below a coarse boundary, while the `freshCommittedTs` is on the next coarse partition? (OK if it does, just checking my thinking here)

Reply:
Yes, it would fail if there is a partition boundary (even fine partition) between the startTs of the first and the second committed write, as we would then only sweep up to the end of the partition.